base/array.jl

```diff
@@ -834,6 +834,43 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
     return A
 end
 
+"""
+    setindex(collection, value, key...)
+
+Create a new collection with the element/elements of the location specified by `key`
+replaced with `value`(s).
+
+# Implementation
+
+`setindex(collection, value, key...)` must have the property such that
+
+```julia
+y1 = setindex(x, value, key...)
+
+y2 = deepcopy(x)
+setindex!(y2, value, key...)
+
+@assert isequal′(y1, y2)
+```
+
+with a suitable definition of `isequal′` (e.g., elements are approximately but possibly
+not exactly equal), _if_ `setindex!` supports given arguments.
+
+`setindex` should support more combinations of arguments by widening collection
+type as required.
+"""
+function setindex end
+
+function setindex(xs::AbstractArray, v, I...)
+    T = promote_type(eltype(xs), typeof(v))
+    ys = similar(xs, T)
+    if eltype(xs) !== Union{}
```

Comment:
In what cases will `promote_type` return `Union{}`?  And why can't we `copy!` a `Union{}` array?

```
julia> copy!(Array{Union{}}(undef, 0), Array{Union{}}(undef, 0))
0-element Array{Union{},1}
```

Reply:
It is sometimes useful to specify the container type and shape without specifying its element type (e.g., for implementing multi-dim `map` in the mutate-or-widen style).  `Union{}` element type is a nice way to encode this. But you can't `copy!` from it.

```julia
julia> copy!(Vector{Int}(undef, 3), Vector{Union{}}(undef, 3))
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] getindex(::Array{Union{},1}, ::Int64) at ./array.jl:758
 [2] copyto!(::Array{Int64,1}, ::Int64, ::Array{Union{},1}, ::Int64, ::Int64) at ./abstractarray.jl:842
 [3] append! at ./array.jl:926 [inlined]
 [4] copy!(::Array{Int64,1}, ::Array{Union{},1}) at ./abstractarray.jl:709
 [5] top-level scope at REPL[5]:1
```

This error itself is not specific to `Union{}`; i.e., `copy!` does not handle undef values.  `eltype(xs) !== Union{}` is just an easy shortcut to handle a useful case.  But maybe this logic should be moved to `copy!` function?

Reply:
I just removed this branching for now. Adding this to `copy!` (in a different PR) seems to be the right direction.